go/adapter.go

```diff
@@ -1,125 +1,67 @@
 package observe
 
 import (
-	"bytes"
-	"errors"
-	"fmt"
-	"math/rand"
-	"time"
+	"context"
+	"log"
 
-	"github.com/tetratelabs/wabin/leb128"
-	"github.com/tetratelabs/wabin/wasm"
+	"github.com/tetratelabs/wazero"
 )
 
+// The primary interface that every Adapter needs to follow
+// Start() and Stop() can just call the implementations on AdapterBase
+// or provide some custom logic. HandleTraceEvent is called after
+// an invocation of a wasm module is done and all events are collected.
 type Adapter interface {
-	Start(collector *Collector, wasm []byte) error
-	Stop(collector *Collector)
-	Event(Event)
+	Start()
+	Stop()
+	HandleTraceEvent(TraceEvent)
 }
 
-type AdapterBase struct {
-	Collectors map[*Collector]chan bool
+// The payload that contains all the Events
+// from a single wasm module invocation
+type TraceEvent struct {
+	Events      []Event
+	TelemetryId *TelemetryId
 }
 
-func checkVersion(m *wasm.Module) error {
-	var minorGlobal *wasm.Export = nil
-	var majorGlobal *wasm.Export = nil
-	for _, export := range m.ExportSection {
-		if export.Type != wasm.ExternTypeGlobal {
-			continue
-		}
-
-		if export.Name == "wasm_instr_version_minor" {
-			minorGlobal = export
-		} else if export.Name == "wasm_instr_version_major" {
-			majorGlobal = export
-		}
-	}
-
-	if minorGlobal == nil || majorGlobal == nil {
-		return errors.New("wasm_instr_version functions not found")
-	}
-
-	minor, _, err := leb128.DecodeUint32(bytes.NewReader(m.GlobalSection[minorGlobal.Index].Init.Data))
-	if err != nil {
-		return err
-	}
-	major, _, err := leb128.DecodeUint32(bytes.NewReader(m.GlobalSection[majorGlobal.Index].Init.Data))
-	if err != nil {
-		return err
-	}
-
-	if major != wasmInstrVersionMajor || minor < wasmInstrVersionMinor {
-		return errors.New(fmt.Sprintf("Expected instrumentation version >= %d.%d but got %d.%d", wasmInstrVersionMajor, wasmInstrVersionMinor, major, minor))
-	}
-
-	return nil
+// Shared implementation for all Adapters
+type AdapterBase struct {
+	TraceEvents chan TraceEvent
+	stop        chan bool
 }
 
-func (a *AdapterBase) Wait(collector *Collector, timeout time.Duration, callback func()) {
-	for {
-		select {
-		case <-time.After(timeout):
-			if len(collector.Events) > 0 {
-				if callback != nil {
-					callback()
-				}
-				continue
-			}
-			a.RemoveCollector(collector)
-			return
-		}
+func (a *AdapterBase) NewTraceCtx(ctx context.Context, r wazero.Runtime, wasm []byte, config *Config) (*TraceCtx, error) {
+	if config == nil {
+		config = NewDefaultConfig()
 	}
+	return newTraceCtx(ctx, a, r, wasm, config)
 }
 
 func NewAdapterBase() AdapterBase {
-	a := AdapterBase{
-		Collectors: map[*Collector]chan bool{},
+	return AdapterBase{
+		// TODO set to some kind of max, add dump logic
+		TraceEvents: make(chan TraceEvent, 100),
```

Comment:
Sends to this channel will block once the channel reaches 100 events. I think thats probably ok, but wondering if this is intentional. We can omit the buffered size and let the channel grow as needed.

Reply:
I see -- yea that makes sense!

Reply:
I'll make a follow up for that. Goal will be to add the bucket on the adapter side and test that this channel can't get blocked. We just need to make sure that the adapter's main routine only reads off this channel as fast as possible. And if it can't put the events in a bucket it should just throw them away and log an error. But never block.

Reply:
SGTM!

Reply:
#44